File: 1. JPA basic/basic/src/main/java/study/jpa/basic/Member.java
package study.jpa.basic;

import javax.persistence.Entity;
import javax.persistence.Id;

/**
 * @author Joonhyuck Hyoung
 * @Entity : JPA가 로딩될때 관리해야 하는 객체로 인식
 * @Id : 최소한 pk가 무엇인지는 알려주어야 한다.
 * @Table(name = "") : 다른 이름의 테이블과 매핑
 * @Column(name = "") : 다른 이름의 컬럼과 매핑
 */

@Entity
public class Member {

    @Id
    private Long id;

    private String name;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

}

File: 5. QueryDSL/querydsl/src/main/kotlin/com/study/querydsl/domain/Member.kt
package com.study.querydsl.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import javax.persistence.*

/**
 * @author Joonhyuck Hyoung
 */
@Entity
class Member(
    @Id
    @GeneratedValue
    @Column(name = "member_id")
    val id: Long = 0L,
    val username: String,
    val age: Int = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id")
    var team: Team? = null,
) {
    fun changeTeam(team: Team?) {
        this.team = team
        team?.members?.add(this)
    }

    override fun toString(): String {
        return "Member(id=${id} username=${username}, age=${age})"
    }

    companion object {
        fun of(username: String) = Member(username = username)

        fun of(username: String, age: Int) = Member(
            username = username,
            age = age
        )
    }
}

